### ISSUE TITLE
Contextual actions and overriding contexts

### ISSUE THREAD

**Issue by FroMage** [eclipse/microprofile-concurrency#115]
Contextual actions are great, but as we can see in #110 sometimes it's not clear what context is going to be applied to an action. Once we contextualise an action, it essentially becomes impossible to change its context, because every other attempts are re-contextualise it will be fruitless.

Take for instance:

```java
void m(Runnable r, ThreadContext context1, ThreadContext context2, ManagedExecutor executor){
 // capture the context here
 Runnable r2 = context1.contextualRunnable(r);

 // Action that changes the context
 ...

 // context2 context operations are no-ops, the current context is ignored
 Runnable r3 = context2.contextualRunnable(r2);
 // executor context operations are no-ops, the current context is ignored
 CompletableFuture<Void> cf4 = executor.runAsync(r2);
}
```

This is because the first contextualised actions created at the last ones to reinstall their context.

I'm not saying this is an issue, but I wonder if it is going to be required at some point to have the ability to re-contextualise or de-contextualise an action in order to change its context or its context settings.

WDYT?

**Comment by njr-11** [comment]
This is a great question.  In general, I would say that the discussion around replacing context might be something to defer to the next spec version.  However, I can also see where current behavior might get in the way of that if we want to be able to reuse the same methods to replace context (which seems like a very natural approach).  To allow for that possibility, it might be wise to state that an attempt to contextualize an interface that is already contextualized will currently raise an error rather than be a no-op.  This would give us the flexibility to add the recontextualization support later.

In other words, we would make the operation context2.contextualRunnable(alreadyContextualizedRunnable)
raise IllegalStateException: Runnable is already contextual.

and then later we would be able to decide if it makes sense to allow that method to be used to replace the context rather than the failure.

**Comment by FroMage** [comment]
Except it's fairly possible that our implementation already does something like this in places, so we may be shooting ourselves in the foot ;)

There's the more convoluted path where using `CompletionStage.thenRun(action)` will contextualise the action, and then later the underlying `CompletableFuture` JDK implementation will create a `Runnable` that it passes to the `ManagedExecutor` which will contextualise it without being able to detect that it already has context because `CompletableFuture` wrapped it into its own `Runnable`. I'm not sure we can detect this case.

**Comment by njr-11** [comment]
> Except it's fairly possible that our implementation already does something like this in places, so we may be shooting ourselves in the foot ;)

I'm also open to allowing for context replacement for contextualRunnable/Callable/Function/... now if that helps.  It would be a pretty trivial change for us, and would basically just require a few tests in the TCK.

Regarding the other problem you mentioned, we completely bypass it by directly supplying a sort of proxy executor to the JDK CompletableFuture where the proxy executor uses the ManagedExecutor for execution but not for context (the context having been previously applied to the action).  That is at least one possible approach here

**Comment by njr-11** [comment]
Here is how I would summarize the rules for overriding context.  If you let me know which way you prefer: IllegalArgumentException vs replacing context, I'll look into updating the JavaDocs accordingly.

threadContext.contextualRunnable(action)
 * Wraps a Runnable with context that is captured from the thread that invokes contextualRunnable _(from existing JavaDoc)_.
 * Raises IllegalArgumentException if an already-contextualized Runnable is supplied to this method.
_OR,_
If an already-contextualized Runnable is supplied to this method, the proxy returned by this method wraps the original Runnable rather than the already-contextualized Runnable proxy.

managedExecutor.supplyAsync/runAsync(action)
 * If an already-contextualized action is supplied, then the action runs with that context.
 * Otherwise, context is captured and/or cleared per the configuration of the ManagedExecutor upon which supplyAsync/runAsync is invoked.

stage.thenRun[Async](action [, executor])
 * If an already-contextualized action is supplied to a completion stage, then the action runs with that context.
 * Otherwise, per outcome of #110

**Comment by njr-11** [comment]
Resolution: where possible, detect collision and raise IllegalArgumentException. For example,
Runnable r = threadContext.contextualRunnable(() -> System.out.println());
Runnable r1 = threadContext.contextualRunnable(r); // can detect this
Runnable r2 = threadContext.contextualRunnable(() -> r.run()); // cannot detect this
